File: src/test/java/com/market/wanted/order/service/OrderServiceTest.java
package com.market.wanted.order.service;

import com.market.wanted.member.entity.Member;
import com.market.wanted.member.repository.MemberRepository;
import com.market.wanted.order.dto.OrderDto;
import com.market.wanted.order.entity.Order;
import com.market.wanted.order.entity.OrderStatus;
import com.market.wanted.product.entity.Product;
import com.market.wanted.product.entity.ProductStatus;
import com.market.wanted.product.repository.ProductRepository;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.annotation.Rollback;
import org.springframework.transaction.annotation.Transactional;

import static org.assertj.core.api.Assertions.*;

@SpringBootTest
@Transactional
@Rollback(false)
class OrderServiceTest {
    @Autowired
    MemberRepository memberRepository;

    @Autowired
    ProductRepository productRepository;

    @Autowired
    OrderService orderService;

    @PersistenceContext
    EntityManager em;

    @Test
    void createOrder() {
        //given
        Member seller = new Member("1234", "seller", "seller@test.com");
        Member buyer = new Member("1234", "buyer", "buyer@test.com");
        memberRepository.save(seller);
        memberRepository.save(buyer);
        Product product = new Product("itemA", 1000, buyer);
        productRepository.save(product);
        //when
        OrderDto order = orderService.createOrder(seller.getUsername(), product.getId());

        //then
        assertThat(order.getBuyerId()).isEqualTo(buyer.getId());
        assertThat(order.getSellerId()).isEqualTo(seller.getId());
        assertThat(order.getOrderStatus()).isEqualTo(OrderStatus.RESERVATION);
    }

    @Test
    void orderConfirm() {
        Member seller = new Member("1234", "seller", "seller@test.com");
        Member buyer = new Member("1234", "buyer", "buyer@test.com");
        memberRepository.save(seller);
        memberRepository.save(buyer);
        Product product = new Product("itemA", 1000, buyer);
        productRepository.save(product);
        OrderDto orderDto = orderService.createOrder(seller.getUsername(), product.getId());

        orderService.orderConfirm(seller.getUsername(), orderDto.getOrderId());

        Order order = orderService.findById(orderDto.getOrderId());

        assertThat(order.getOrderStatus()).isEqualTo(OrderStatus.COMPLETE);
        assertThat(order.getOrderItem().getProduct().getStatus()).isEqualTo(ProductStatus.SOLD_OUT);
    }

}

File: src/test/kotlin/wanted/market/api/order/application/OrderServiceTest.kt
package wanted.market.api.order.application

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import jakarta.persistence.EntityManager
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional
import wanted.market.api.member.application.MemberAuthService
import wanted.market.api.member.domain.dto.`in`.CommandRegisterMember
import wanted.market.api.member.domain.dto.out.CommandRegisterMemberResult
import wanted.market.api.order.domain.entity.OrderStatus.SALEAPPROVAL
import wanted.market.api.product.application.ProductCommandService
import wanted.market.api.product.domain.dto.`in`.CommandPurchaseProduct
import wanted.market.api.product.domain.dto.`in`.CommandRegisterProduct
import wanted.market.api.product.domain.dto.out.CommandRegisterProductResult
import wanted.market.api.product.repository.ProductRepository

@SpringBootTest
@Transactional
class OrderServiceTest(
    private val orderService: OrderService,
    private val memberAuthService: MemberAuthService,
    private val productCommandService: ProductCommandService,
    private val em: EntityManager,
    private val productRepository: ProductRepository
) : StringSpec({

    fun registerMember(): CommandRegisterMemberResult {
        val request = CommandRegisterMember(
            userId = "seonwoo_jung",
            password = "12345678a"
        )

        // when
        return memberAuthService.registerMember(request)
    }

    fun registerProduct() : CommandRegisterProductResult {
        val memberInfo = registerMember()

        val productRequest = CommandRegisterProduct(
            name = "상품명",
            price = 1_000_000,
            quantity = 2
        )
        return productCommandService.registerProduct(productRequest, memberInfo.id)
    }

    "판매자가 제품 판매 승인을 한다" {
        val product = registerProduct()

        val purchaseRequest = CommandPurchaseProduct(
            productId = product.id,
            quantity = 2
        )

        // when
        val purchaseResponse = productCommandService.purchaseProduct(purchaseRequest, product.sellerId)

        val order = orderService.approveOrder(purchaseResponse.id, product.sellerId)

        order.orderStatus shouldBe SALEAPPROVAL
        em.flush()
    }

    "구매자가 구매 확정을 한다" {
        val product = registerProduct()

        val purchaseRequest = CommandPurchaseProduct(
            productId = product.id,
            quantity = 2
        )

        // when
        val purchaseResponse = productCommandService.purchaseProduct(purchaseRequest, product.sellerId)

        val order = orderService.approveOrder(purchaseResponse.id, product.sellerId)

        em.flush()

        orderService.confirmOrder(order.orderId, product.sellerId)
        em.flush()
    }
})
